### PR TITLE
[8.x] Eager Loaded Pseudo Relationships

### PR DESCRIPTION
**Pull request by browner12**
### TLDR

Create pseudo relationship that can be eager loaded:

```php
$salesperson = Salesperson::with(['leads as openLeads' => fn($query) => $query->where('status', 'open')])
    ->with(['leads as completedLeads' => fn($query) => $query->where('status', 'completed')])
    ->with(['leads as completedLeadsWithSale' => fn($query) => $query->where('status', 'completed')->where('result', 'sale')])
    ->with(['leads as completedLeadsWithNoSale' => fn($query) => $query->where('status', 'completed')->where('result', 'no sale')])
    ->first();

$salesperson->openLeads;
$salesperson->completedLeads;
$salesperson->completedLeadsWithSale;
$salesperson->completedLeadsWithNoSale;
```

---

### Current Scenario

Imagine you have a `Salesperson` with many `Lead`s. `Lead`s have a `status` of "open" or "completed", and a `result` of "sale" or "no sale".

You are trying to build a dashboard that shows the `Salesperson`'s metrics.

| Salesperson | Open | Completed | No Sale | Sale | Close Rate |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Taylor  | 15 | 20 | 15 | 5 | 25% |
| Mohammed  | 7 | 6 | 1 | 5 | 83% |
| Dries  | 12 | 15 | 5 | 10 | 66% |

It is a simple 1:many relationship between the models.

```php
class Salesperson extends Model
{
    public function leads()
    {
        return $this->hasMany(Lead::class);
    }
}
```

In your controller you fetch all the `Salesperson`s.

```php
class SalespersonController
{
    public function index()
    {
        $salespersons = Salesperson::with('leads')->get();
    }
}
```

In your view you loop through the `Salesperson`s.

```blade
@foreach($salespersons as $salesperson)
    <tr>
        <td>{{ $salesperson->name }}</td>
        <td>{{ count($salesperson->leads()->where('status', 'open')->get()) }}</td>
        <td>{{ count($salesperson->leads()->where('status', 'completed')->get()) }}</td>
        <td>{{ count($salesperson->leads()->where('status', 'completed')->where('result', 'no sale')->get()) }}</td>
        <td>{{ count($salesperson->leads()->where('status', 'completed')->where('result', 'sale')->get()) }}</td>
        <td>{{ count($salesperson->leads()->where('status', 'completed')->where('result', 'sale')->get()) / count($salesperson->leads()->where('status', 'completed')->get()) }}</td>
    </tr>
@endforeach
```

Because we are calling `->leads()` as a method and not a property (in order to tack on our additional conditions), this will result in new queries being run for each of these groupings, and we don't benefit from the eager loading we did in the controller.

There are 2 ways we could currently handle this. We could create additional constrained relationships in the Model:

```php
class Salesperson extends Model
{
    public function leads()
    {
        return $this->hasMany(Lead::class);
    }

    public function openLeads()
    {
        return $this->hasMany(Lead::class)->where('status', 'open');
    }

    public function completedLeads()
    {
        return $this->hasMany(Lead::class)->where('status', 'completed');
    }
    
    public function completedLeadsWithSale()
    {
        return $this->hasMany(Lead::class)->where('status', 'completed')->where('result', 'sale');
    }
    
    public function completedLeadsWithNoSale()
    {
        return $this->hasMany(Lead::class)->where('status', 'completed')->where('result', 'no sale');
    }
}
```

and then eager load these in the Controller:

```php
class SalespersonController
{
    public function index()
    {
        $salespersons = Salesperson::with('leads')
            ->with('openLeads')
            ->with('completedLeads')
            ->with('completedLeadsWithSale')
            ->with('completedLeadsWithNoSale')
            ->get();
    }
}
```

This gets rid of our N+1, but the problem with this is it quickly becomes very verbose and unmanageable.

The other option is to filter in PHP (rather than the query) using the returned Collections. In our controller we revert to our simple eager load on the `leads` relationship:

```php
class SalespersonController
{
    public function index()
    {
        $salespersons = Salesperson::with('leads')->get();
    }
}
```

and now in our view we use Collection methods to filter out our desired data:

```blade
@foreach($salespersons as $salesperson)
    <tr>
        <td>{{ $salesperson->name }}</td>
        <td>{{ count($salesperson->leads->where('status', 'open')) }}</td>
        <td>{{ count($salesperson->leads->where('status', 'completed')) }}</td>
        <td>{{ count($salesperson->leads->where('status', 'completed')->where('result', 'no sale')) }}</td>
        <td>{{ count($salesperson->leads->where('status', 'completed')->where('result', 'sale')) }}</td>
        <td>{{ count($salesperson->leads->where('status', 'completed')->where('result', 'sale')) / count($salesperson->leads()->where('status', 'completed')) }}</td>
    </tr>
@endforeach
```

This option benefits from the eager loading, so our query count is very low. However, there is a lot of "logic" in the view, which some people do not like. We also duplicate 2 conditions in our "Close Rate" column that we've already determined. Finally, this does offload the filtering and sorting to PHP rather than the database, which could be undesirable in some scenarios, possibly for performance reasons.

---

### Proposed Improvement

This PR allows us to define eager loaded pseudo relationships. The benefit to this is it avoids our N+1 issues, keeps the filtering and sorting in the database, and keeps the majority of the logic out of the view.

```php
class SalespersonController
{
    public function index()
    {
        $salespersons = Salesperson::with(['leads as openLeads' => fn($query) => $query->where('status', 'open')])
            ->with(['leads as completedLeads' => fn($query) => $query->where('status', 'completed')])
            ->with(['leads as completedLeadsWithSale' => fn($query) => $query->where('status', 'completed')->where('result', 'sale')])
            ->with(['leads as completedLeadsWithNoSale' => fn($query) => $query->where('status', 'completed')->where('result', 'no sale')])
            ->get();
    }
}
```

These "relationships" are then available on the models in the view:

```blade
@foreach($salespersons as $salesperson)
    <tr>
        <td>{{ $salesperson->name }}</td>
        <td>{{ count($salesperson->openLeads) }}</td>
        <td>{{ count($salesperson->completedLeads) }}</td>
        <td>{{ count($salesperson->completedLeadsWithNoSale) }}</td>
        <td>{{ count($salesperson->completedLeadsWithSale) }}</td>
        <td>{{ count($salesperson->completedLeadsWithSale) / count($salesperson->completedLeads) }}</td>
    </tr>
@endforeach
```

Thanks!

### CODE REVIEW

**Comment by taylorotwell** [comment]
In your examples you pass an array to `with` that has two elements... a string with a relationship and an alias and a second value that is a closure. Is that correct? Is it not supposed to be like:

```php
::with(['leads as openLeads' => fn ($query) ... ])
```

**Comment by browner12** [comment]
You are correct. Tried coding freehand in the markdown, clearly not a good idea.

Comment has been updated.

**Comment by fredsal** [comment]
Maybe a duplicate of [PR #31976](https://github.com/laravel/framework/pull/31976)

**Comment by fredsal** [comment]
Works on nested eager loading `with('author.contacts as contacts_data')`?
Like this branch
https://github.com/CrissiDev/framework/blob/f8004c930bf64055e9687c85356195e0e04304d3/tests/Database/DatabaseEloquentIntegrationTest.php#L841-L857

**Comment by PaolaRuby** [comment]
Can you take a look on this?
https://github.com/laravel/framework/blob/09feffba0ba3ebff4a366f020519b34c3d2e101f/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php#L368-L379

**Comment by donnysim** [comment]
We should not forget about 
```php
->with(['leads:id,name')
```
as in will this:
```php
->with(['leads:id,name as completedLeads')
```
still work?

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by PaolaRuby** [comment]
@taylorotwell @donnysim already works with `leads:id,name as autor`
```php
public function testBasicEagerLoadingWithAlias()
{
    $user = EloquentTestUser::create(['email' => 'taylorotwell@gmail.com']);
    $user->posts()->create(['name' => 'First Post']);
    $user->posts()->create(['name' => 'Second Post']);
    $user = EloquentTestUser::with([
        'posts:id,user_id,name AS autor',
        'posts AS posts1:id,user_id,name AS autor,parent_id',
        'posts AS posts2' => function ($q) {
            $q->where('name', 'Second Post');
        },
    ])->where('email', 'taylorotwell@gmail.com')->first();
    $this->assertSame('First Post', $user->posts->first()->autor);
    $this->assertSame('First Post', $user->posts1->first()->autor);
    $this->assertSame('Second Post', $user->posts2->first()->name);
}
```
`.\vendor\bin\phpunit --filter=testBasicEagerLoadingWithAlias`
```
PHPUnit 9.5.6 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.0.5
Configuration: laravel/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 00:00.088, Memory: 126.00 MB

OK (1 test, 3 assertions)
```

**Comment by Lemmings19** [comment]
Different PR for the same feature a little over a year ago, same outcome/response. :(

https://github.com/laravel/framework/pull/31976#issuecomment-599538731

**Comment by OzanKurt** [comment]
Did anybody care to extract this funtionality to a package yet?

**Comment by angeljqv** [comment]
@OzanKurt no, feel free to make it possible

**Comment by shijunti19** [comment]
When I use the attachment ID to save the attachment and then read the attachment, it is very useful. I only need to obtain a layer of array

**Comment by shinsenter** [comment]
Will there be any updates in 2024?
